### PR TITLE
rst_deck: application to create a deck prepared for restart

### PR DESCRIPTION
**Pull request by joakim-hove**
A small application to create a restart deck. This is a first implementation - it is usable, but there are indeed plans for improvements. The two main missing features are:

1. The argument used to denote what to restart from e.g. `base/HISTORY:60` is only interpreted *as a string*, i.e. we do no file system shuffling on this argument, we do not check/update with `UNIFIN/MULTIN` or `FMTIN` and we do not check the contents of the restart file.
2. For Schedule maniupulations we *only* support adding a `SKIPREST` after the `SCHEDULE` keyword, and not removing keywords up until the correct dates.

Both 1 and 2 above should be feasible, but I wanted to get something merged now. The `rst_deck -h` is showed below:
```
The rst_deck program will load a simulation deck and parameters for a restart
and reformat the deck to become a restart deck. Before the updated deck is
output the program will update the SOLUTION and SCHEDULE sections. All keywords
from the SOLUTION section will be cleared out(1) and a RESTART keyword will be
inserted. In the SCHEDULE section the program can either remove all keywords up
until the restart date, or alternatively insert SKIPREST immediately following
the SCHEDULE keyword(2).

When creating the updated restart deck the program can either link to unmodified
include files with INCLUDE statements, create a copy of deck structure in an
alternative location or create one large file with all keywords in the same
file. Apart from the alterations to support restart the output deck will be
equivalent to the input deck, but formatting is not retained and comments have
been stripped away.

Arguments:

1. The data file we are starting with.

2. The basename of the restart file - with an optional path prefix and a :N to
   restart from step N(3). A restart step value of 0 is interpreted as a dry run
   - a deck which has not been set up for restart will be written out.

3. Basename of the restart deck we create, can optionally contain a path prefix;
   the path will be created if it does not already exist. This argument is
   optional, if it is not provided the program will dump a restart deck on
   stdout. If the argument corresponds to an existing directory the restart case
   will get the same name as the base case.

Options:

-s: Manipulate the SCHEDULE section by inserting a SKIPREST keyword immediately
    following the SCHEDULE keyword. If the -s option is not used the SCHEDULE
    section will be modified by removing all keywords until we reach the restart
    date. NB: Currently the -s option is required

-m: [share|inline|copy] The restart deck can reuse the unmodified include files
    from the base case, this is mode 'share' and is the default. With mode
    'inline' the restart deck will be one long file and with mode 'copy' the
    file structure of the base case will be retained. The default if no -m
    option is given is the 'share' mode.

Example:

   rst_deck /path/to/history/HISTORY.DATA rst/HISTORY:30 /path/to/rst/RESTART -s

1: The program has a compiled list of keywords which will be retained in the
   SOLUTION section. The current value of that list is: {}

2: Current version of the program *only* supports the SKIPREST option, and the
   -s option is required.

3: The second argument is treated purely as a string and inserted verbatim into
   the updated restart deck. In a future version we might interpret the second
   argument as a file path and check the content and also do filesystem
   manipulations from it.
```

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
@akva2 : This has been long in the coming, but is is now approaching something which can be merged and used (as a first step ...). The last commit adds some integration testing of the functionality - I have made an absolute minimal "Works for me" approach to integrating the test - but it is probably not up to your standards? At the very least the dependencies on the executables `rst_deck` and `opmhash` is not expressed. A little help to get by?

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
> A little help to get by?

I would like to help, but there is very little info here, not even a PR description. What is this?

**Comment by joakim-hove** [comment]
> I would like to help, but there is very little info here, not even a PR description. What is this?

Hehe - thank you; PR description is coming. The plea for some assistance is concerned with the testing in the last commit - described in the [comment](https://github.com/OPM/opm-common/pull/2684#issuecomment-925105188).

File: ExtraTests.cmake [hunk 1]
```diff
@@ -69,6 +69,11 @@ opm_add_test( rst_msw
               LIBRARIES ${TEST_LIBS}
               TEST_ARGS tests/MSW.DATA tests/MSW_RESTART.DATA )
 
+add_test( NAME rst_deck_test
+          COMMAND ${PROJECT_SOURCE_DIR}/tests/rst_test_driver.sh ${CMAKE_BINARY_DIR}/bin/rst_deck ${CMAKE_BINARY_DIR}/bin/opmhash
```
Comment (akva2) [inline ExtraTests.cmake:73]: PROJECT_BINARY_DIR

File: ExtraTests.cmake [hunk 2]
```diff
@@ -113,8 +118,17 @@ if(HAVE_OPM_TESTS)
   opm_add_test("SPE9_CP_GROUP2" NO_COMPILE EXE_NAME parse_write TEST_ARGS "${OPM_TESTS_ROOT}/spe9group/SPE9_CP_GROUP.DATA")
   set_property(TEST NORNE_ATW2013
                PROPERTY ENVIRONMENT "OPM_ERRORS_IGNORE=PARSE_RANDOM_SLASH")
+
+  add_test( NAME rst_deck_test_norne
+             COMMAND ${PROJECT_SOURCE_DIR}/tests/rst_test_driver.sh ${CMAKE_BINARY_DIR}/bin/rst_deck ${CMAKE_BINARY_DIR}/bin/opmhash
```
Comment (akva2) [inline ExtraTests.cmake:123]: PROJECT_BINARY_DIR

**Comment by akva2** [comment]
it seems to work but i have to manually add UNIFIN, probably something you missed?

File: ExtraTests.cmake (hunk 1 above)
Comment (akva2) [reply ExtraTests.cmake:73]: test execution is separate from building, and you cannot add dependencies to the test target in the build system. the best you can do is to make sure it works with our custom 'check' target (which builds, and then executes tests). this is done automatically for anything that is built through EXAMPLE_SOURCE_FILES, so there is nothing more to be done.

**Comment by joakim-hove** [comment]
> it seems to work but i have to manually add UNIFIN, probably something you missed?

I'll look into that.

File: ExtraTests.cmake (hunk 1 above)
Comment (akva2) [reply ExtraTests.cmake:73]: however, we would want the path of rst_deck to be exposed to downstreams where we will use it. you do this like we do for OPM_PACK on COMPARE_ECL - see https://github.com/OPM/opm-common/blob/master/CMakeLists.txt#L43 and the block below

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
I don't want to be snide, but there's no way this code compiled on your own machine.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
> I don't want to be snide, but there's no way this code compiled on your own machine.

Believe it or not - it did compile on my machine - but I had turned off the cmake switch `BUILD_EXAMPLES`.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
> it seems to work but i have to manually add UNIFIN, probably something you missed?

@akva2 : I think this PR should be merged in approximately the current state. That implies that updates to e.g. UNIFIN must still be performed manually, I agree that e.g. adding a UNIFIN is within the scope of this work, but it is not obvious (from a domain perspective) exactly how that should be done - will therefor relegate such issues to a subsequent PR.

**Comment by akva2** [comment]
fair enough. i can't update the regtest support to use this before UNFIN is in but that is not a reason not to merge this in bite sizes.

**Comment by joakim-hove** [comment]
> i can't update the regtest support to use this before UNFIN is

Well - I don't really agree on that - we (or I) - can just add and check in `UNIFIN` to the test decks in opm-tests, that does not affect the running of the normal regression tests.

**Comment by akva2** [comment]
ah okay, it would dump it out again, i see. yeah then we can do that.

**Comment by joakim-hove** [comment]
> yeah then we can do that.

OK - I'm scripting it; PR coming.

**Comment by joakim-hove** [comment]
Opinions on this?

**Comment by joakim-hove** [comment]
jenkins build this please

**Pull request merged**